source/workbook/workbook.cpp

```diff
@@ -1563,69 +1577,83 @@ void workbook::update_sheet_properties()
     }
 }
 
-void workbook::reorder_relationships()
+namespace {
+// true if a sheet index is != worksheet relationship index
+bool needs_reorder(const std::unordered_map<std::string, std::string> &title_to_rels,
+    const std::vector<std::string> &titles,
+    std::vector<std::string> &relation_ids)
 {
-    const auto wb_rel = manifest().relationship(path("/"), relationship_type::office_document);
-    const auto wb_path = wb_rel.target().path();
-    const auto relationships = manifest().relationships(wb_path);
-    std::unordered_map<std::string, relationship> rel_map;
-    const auto titles = sheet_titles();
-    const auto title_rel_id_map = d_->sheet_title_rel_id_map_;
-    bool needs_reorder = false;
-
-    for (const auto &rel : relationships)
+    bool all_match = true;
+    for (std::size_t title_index = 0; title_index < titles.size(); ++title_index)
     {
-        rel_map[rel.id()] = rel;
-
-        if (rel.type() == relationship_type::worksheet)
+        const auto &rel = title_to_rels.at(titles[title_index]);
+        relation_ids.push_back(rel);
+        const auto expected_rel_id = "rId" + std::to_string(title_index + 1);
+        if (rel != expected_rel_id)
         {
-            for (auto title_index = std::size_t(0); title_index < titles.size(); ++title_index)
-            {
-                const auto title = titles[title_index];
-
-                if (title_rel_id_map.at(title) == rel.id())
-                {
-                    const auto expected_rel_id = "rId" + std::to_string(title_index + 1);
-
-                    if (expected_rel_id != rel.id())
-                    {
-                        needs_reorder = true;
-                    }
+            all_match = false;
+        }
+    }
+    return !all_match; // if all are as expected, reorder not required
+};
 
-                    break;
-                }
-            }
+struct rel_id_sorter
+{
+    // true if lhs < rhs
+    bool operator()(const xlnt::relationship &lhs, const xlnt::relationship &rhs)
+    {
+        // format is rTd<decimal number 1..n>
+        if (lhs.id().size() < rhs.id().size()) // a number with more digits will be larger
+        {
+            return true;
         }
+        return lhs.id() < rhs.id();
     }
+};
+} // namespace
 
-    if (!needs_reorder)
+void workbook::reorder_relationships()
```

Comment:
While this implementation resolves #279, the complexity here is very high. #316 created as a starting point for simplifying relationship management